usr/src/uts/common/io/overlay/overlay_router.c

```diff
@@ -0,0 +1,2389 @@
+
+	if (avl_find(&orr->orr_nets_vlan, net, NULL) != NULL ||
+	    avl_find(&orr->orr_nets_mac, net, NULL) != NULL ||
+	    (hasv4 && avl_find(&orr->orr_nets_v4, net, NULL) != NULL) ||
+	    (hasv6 && avl_find(&orr->orr_nets_v6, net, NULL) != NULL)) {
+		mutex_exit(&net->ont_lock);
+		mutex_exit(&orr->orr_lock);
+
+		overlay_net_rele(net);
+		return (EEXIST);
+	}
+
+	avl_add(&orr->orr_nets_vlan, net);
+	net->ont_refcnt++;
+
+	avl_add(&orr->orr_nets_mac, net);
+	net->ont_refcnt++;
+
+	if (hasv4) {
+		avl_add(&orr->orr_nets_v4, net);
+		net->ont_refcnt++;
+	}
+
+	if (hasv6) {
+		avl_add(&orr->orr_nets_v6, net);
+		net->ont_refcnt++;
+	}
+
+	mutex_exit(&net->ont_lock);
+	mutex_exit(&orr->orr_lock);
+
+	/* Release the initial hold we took immediately after allocation. */
+	overlay_net_rele(net);
+	return (0);
+}
+
+static int
+overlay_router_net_delete(overlay_router_t *orr, void *buf)
+{
+	overlay_ioc_net_t *rnet = buf;
+	overlay_net_t *net = NULL;
+
+	/*
+	 * Require all other fields aside from the overlay id and the
+	 * router id to be zero. We want to prevent a scenario where
+	 * current clients pass in garbage for currently unused fields
+	 * (and have it work) -- if other fields are used in the future this
+	 * could be a source of interesting bugs.
+	 */
+	if (!IN6_IS_ADDR_UNSPECIFIED(&rnet->oin_routeraddrv6) ||
+	    rnet->oin_routeraddr != INADDR_ANY ||
+	    rnet->oin_prefixlen != 0 ||
+	    rnet->oin_prefixlenv6 != 0 ||
+	    rnet->oin_vlan != 0 ||
+	    bcmp(rnet->oin_mac, overlay_macaddr, ETHERADDRL) != 0 ||
+	    strlen(rnet->oin_routetbl) != 0)
+		return (EINVAL);
+
+	if (!overlay_valid_id(rnet->oin_id, sizeof (rnet->oin_id)))
+		return (EINVAL);
+
+	if (strlen(rnet->oin_id) == 0)
+		return (EINVAL);
+
+	net = overlay_hold_net_by_id(orr, rnet->oin_id);
+	if (net == NULL)
+		return (ENOENT);
+
+	mutex_enter(&orr->orr_lock);
+	mutex_enter(&net->ont_lock);
+
+	avl_remove(&orr->orr_nets_vlan, net);
+	net->ont_refcnt--;
+
+	avl_remove(&orr->orr_nets_mac, net);
+	net->ont_refcnt--;
+
+	if (net->ont_routeraddr != INADDR_ANY) {
+		avl_remove(&orr->orr_nets_v4, net);
+		net->ont_refcnt--;
+	}
+
+	if (!IN6_IS_ADDR_UNSPECIFIED(&net->ont_routeraddrv6)) {
+		avl_remove(&orr->orr_nets_v6, net);
+		net->ont_refcnt--;
+	}
+	mutex_exit(&orr->orr_lock);
+	mutex_exit(&net->ont_lock);
+
+	overlay_net_rele(net);
+	return (0);
+}
+
+static int
+overlay_router_net_delete_all(overlay_router_t *orr, void *buf)
+{
+	overlay_ioc_net_t *rnet = buf;
+
+	/*
+	 * Similar to overlay_router_delete(), we enforce that only
+	 * the overlay id is set, and all other fields are 0.
+	 */
+	if (!IN6_IS_ADDR_UNSPECIFIED(&rnet->oin_routeraddrv6) ||
+	    rnet->oin_routeraddr != INADDR_ANY ||
+	    rnet->oin_prefixlen != 0 ||
+	    rnet->oin_prefixlenv6 != 0 ||
+	    rnet->oin_vlan != 0 ||
+	    bcmp(rnet->oin_mac, overlay_macaddr, ETHERADDRL) != 0 ||
+	    strlen(rnet->oin_id) != 0 ||
+	    strlen(rnet->oin_routetbl) != 0)
+		return (EINVAL);
+
+	/* TODO */
```

Comment:
XXX KEBE ASKS what's the plan for this?

Reply:
The idea is to add a 'flush' command, similar to `route flush`. Just ran out of time.

Reply:
Ack.  Please note that?